### PR TITLE
τ_z dedicated subnet — attack residual axis (test_τ_z ≈ 9.05% across no-SDF runs)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Add a dedicated τ_z prediction subnet that runs in parallel with the existing shared surface head, attacking the residual axis directly.**

Three independent no-SDF tay runs at full convergence now all show test_τ_z ≈ 9.05% as the worst axis:

| Run | test_τ_x | test_τ_y | **test_τ_z** | test_WSS |
|---|---:|---:|---:|---:|
| alphonse #1078 (asymmetric eval) | 6.190% | 7.635% | **9.073%** | 6.996% |
| **thorfinn #1100 (slices=256)** | **6.208%** | **7.581%** | **9.051%** | **6.989%** |
| PR #972 SOTA (SDF stack) | — | — | (lower) | 6.727% |

τ_z is **18bp worse on average than τ_y** (the second-worst axis) and **29bp worse than τ_x**. This is consistent program-wide across capacity-uplift, eval-resolution-uplift, and the baseline. The Wave 27 finding ("91-96% of WSS residual is magnitude error") combined with this per-axis pattern strongly suggests τ_z magnitude prediction is the single biggest available improvement to test_WSS.

### Why a dedicated subnet (and not just per-channel output heads)

Edward #1116 is testing per-channel **output heads** — decoupling the final linear projection so τ_x/τ_y/τ_z each get their own linear layer over the shared hidden representation. That tests whether the OUTPUT linear is the bottleneck.

This experiment tests whether the **representation** going into the τ_z output is the bottleneck. A dedicated subnet:
- Has its own 2-layer MLP (with non-linearity) sitting between the surface hidden state and the τ_z output
- Allows the model to learn τ_z-specific intermediate features rather than relying on features shared with τ_x/τ_y/SP/SDF
- Compounds with edward's per-channel heads if both are additive (could be tested in a future combined run)

If τ_z's residual error is structurally driven by mid-layer representations being optimized for τ_x/τ_y (which are easier and have larger gradient magnitudes during training), a dedicated subnet with its own capacity will get its own gradient signal isolated from that competition.

### Connection to Wave 27 magnitude finding

Wave 27 established that direction is essentially solved (cos_sim=0.996 by EP2). The residual is magnitude error, concentrated on τ_z. A dedicated subnet for τ_z is the architecturally minimal way to test if **per-axis magnitude prediction benefits from per-axis representational capacity**.

### Implementation in `model.py` — add τ_z dedicated subnet

In the `SurfaceTransolver.__init__()`, after the existing `self.surface_out` block (around line 484-499), add a conditional dedicated τ_z subnet:

```python
# In __init__, after existing self.surface_out block:
self.use_tau_z_subnet = bool(use_tau_z_subnet)
if self.use_tau_z_subnet:
    subnet_hidden = int(n_hidden * tau_z_subnet_hidden_mult)
    layers = []
    in_dim = n_hidden
    for _ in range(tau_z_subnet_depth):
        layers.append(nn.Linear(in_dim, subnet_hidden))
        layers.append(nn.GELU())
        in_dim = subnet_hidden
    layers.append(nn.Linear(in_dim, 1))  # τ_z output
    self.tau_z_subnet = nn.Sequential(*layers)
    self.tau_z_subnet.apply(_init_linear)
```

Add the constructor params:
```python
def __init__(
    self,
    ...
    use_tau_z_subnet: bool = False,
    tau_z_subnet_depth: int = 2,
    tau_z_subnet_hidden_mult: float = 1.0,
    ...
):
```

In `forward()`, after computing `surface_preds = self.surface_out(surface_hidden) * surface_mask.unsqueeze(-1)` (around line 625), add:

```python
if self.use_tau_z_subnet:
    tau_z_dedicated = self.tau_z_subnet(surface_hidden) * surface_mask.unsqueeze(-1)
    # Replace τ_z slice of surface_preds with dedicated prediction
    # surface_preds layout: [SP, tau_x, tau_y, tau_z] = indices [0, 1, 2, 3]
    surface_preds = surface_preds.clone()
    surface_preds[..., 3:4] = tau_z_dedicated
```

**Verify the τ_z index is 3** — the SURFACE_Y_DIM=4 layout in this codebase should be [SP, τ_x, τ_y, τ_z]. Grep `SURFACE_Y` and confirm before running.

### CLI flags in `train.py`

Add to `Config`:
```python
use_tau_z_subnet: bool = False
tau_z_subnet_depth: int = 2
tau_z_subnet_hidden_mult: float = 1.0
```

Add to argparse and pass through to `build_model()`.

### Optional: dedicated loss weight (NOT required for first run)

The shared head already produces a τ_z prediction. With the subnet active, you have two options:
- **Option A (recommended for first run)**: Replace shared head's τ_z with the dedicated subnet's output entirely (as in the forward modification above). Loss is unchanged — same MSE on the [SP, τ_x, τ_y, τ_z_dedicated] surface_preds vector.
- **Option B**: Keep both, add a supplementary loss `λ_tau_z_aux * MSE(tau_z_shared, tau_z_gt)` where the shared head's τ_z is supervised as an auxiliary signal. Adds complexity; defer for v2 if v1 is positive.

Stick with Option A. The hypothesis is "dedicated representational capacity for τ_z helps" — A is the cleanest test.

### Smoke test (1-2 epochs to confirm subnet is wired correctly)

```bash
cd target/ && SENPAI_TIMEOUT_MINUTES=270 torchrun --standalone --nproc-per-node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --use-ema --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 2 \
  --vol-points-schedule "0:16384" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --use-tau-z-subnet --tau-z-subnet-depth 2 --tau-z-subnet-hidden-mult 1.0 \
  --wandb_group tau-z-dedicated-subnet --agent thorfinn \
  --wandb_name "thorfinn/tau-z-subnet-smoke"
```

Smoke PASS:
- 2 epochs complete cleanly (no NaN, no shape mismatch crash)
- Param count in W&B has increased by `~(n_hidden*n_hidden*depth + n_hidden)` params (for default depth=2, mult=1.0, n_hidden=512: +786k params, modest)
- EP1 val_abupt ≤ 35% (gentle catastrophe gate with lr-warmup-epochs=1 ~33% baseline)
- EP2 val_abupt should be in range with baseline frieren `mempfubx` EP2 numbers (no regression from adding params)

### Full 13-ep run (DDP 8 GPUs, 18h budget)

```bash
cd target/ && SENPAI_TIMEOUT_MINUTES=1100 nohup torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --use-ema --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --use-tau-z-subnet --tau-z-subnet-depth 2 --tau-z-subnet-hidden-mult 1.0 \
  --wandb_group tau-z-dedicated-subnet --agent thorfinn \
  --wandb_name "thorfinn/tau-z-subnet-full13ep" \
  > logs/tau-z-subnet.log 2>&1 &
```

### EP3 mid-flight gate

- val_abupt ≤ 7.2% PASS / ≤ 7.6% MARGINAL / > 7.6% KILL
- val_vol_p ≤ 4.5% PASS / ≤ 5.0% MARGINAL / > 5.0% KILL

If PASS, continue. If KILL, post EP3 metrics + confirmation that `use_tau_z_subnet=True` is in W&B config — sometimes architectural changes silently no-op when the flag plumbing is incomplete.

### EP3 specifically: τ_z / τ_x ratio diagnostic

The headline diagnostic for this experiment: **val_τ_z / val_τ_x at EP3**. Anchor values from past runs:
- Baseline (no subnet): val_τ_z / val_τ_x at EP3 ≈ 1.6-1.7 (from previous Wave 28 EP3 numbers)
- Target: val_τ_z / val_τ_x at EP3 < 1.5 — meaningful reduction
- Strong signal: val_τ_z / val_τ_x at EP3 < 1.4 — dedicated subnet is clearly attacking the bottleneck
- Null result: val_τ_z / val_τ_x at EP3 ≈ 1.6-1.7 — subnet adds params but doesn't help τ_z specifically

### EP13 final + auto-harvest

After EP13 (~13-14h wall-clock at validated 18h budget cadence), auto-harvest runs test eval from best-val EMA checkpoint.

**Merge gate**:
- test_WSS < 6.989% (beats no-SDF ceiling identified by alphonse #1078 + thorfinn #1100) AND test_vol_p ≤ 3.643% AND test_SP ≤ 3.577%
- If test_τ_z drops below 9.0%, that alone is paper-relevant even if test_WSS doesn't beat SOTA

**Consider gate**:
- Floors held AND test_WSS in [6.989%, 7.05%]
- Direction: "τ_z subnet is methodology-positive but capacity-bound; try larger subnet or combine with SDF"

**Close gate**:
- Test floors regress (subnet would have to silently corrupt training, surprising)
- val_abupt > 7.6% at EP3 KILL

### Diagnostic to report at terminal SENPAI-RESULT

In addition to standard test/val metrics:

1. **Per-axis test breakdown** — test_WSSx, test_WSSy, **test_WSSz** vs the baseline 9.051% (thorfinn #1100) / 9.073% (alphonse #1078). The headline number is test_WSSz.
2. **τ_z / τ_x ratio** at terminal — should be 1.46 (thorfinn baseline) → ideally < 1.35 if subnet works.
3. **Param count delta** vs baseline thorfinn #1100 run — at depth=2 mult=1.0, this should be +~786k params (small relative to model total).
4. **VRAM ceiling** — at slices=128 you have plenty of room; this should land comfortably under 80GB.
5. **Compare to edward #1116 per-channel-heads result** when it lands — your subnet attacks representational capacity; his attacks output capacity. If both improve τ_z, the combined experiment is the obvious Wave 29 follow-up.

## Baseline

Current best (BASELINE.md, 2026-05-14):
- val_abupt: 6.126% (PR #972 SDF stack SOTA val)
- val_WSS: 7.155% (PR #972)
- **test_WSS: 6.727% (PR #972 single-model SOTA, paper-facing)**
- test_vol_p: 3.643% (floor — MUST NOT regress)
- test_SP: 3.577% (floor — MUST NOT regress)

**No-SDF tay structural ceiling** (alphonse #1078 + thorfinn #1100 convergence):
- test_WSS ≈ 6.99% — capacity-uplift cannot break this without SDF
- test_τ_z ≈ 9.05% — the residual axis your subnet targets

Live competing runs:
- alphonse #1122: SDF importance sampling port to tay (unrelated, complementary)
- edward #1116: per-channel WSS output heads (related but tests output capacity not subnet)
- fern #1119: GradNorm short-cycle (loss-side mechanism)
- nezuko #1120: spatial-prior surface sampling
- frieren #1121: WSS magnitude-only decomposition + 18h budget
- askeladd #1118: OHEM v2 spike-clipped
- tanjiro #1114: learnable WSS channel weights

## Decision criteria

**MERGE** if all three:
- test_WSS < 6.989% (beat no-SDF ceiling) AND test_vol_p ≤ 3.643% AND test_SP ≤ 3.577%

**CONSIDER**:
- Floors held AND test_WSS in [6.989%, 7.05%] — subnet methodology-positive but not enough alone
- test_τ_z < 9.0% standalone — paper-relevant even without WSS win

**CLOSE**:
- Floors regress (surprising for a +786k-param subnet)
- val_abupt > 7.6% at EP3 KILL

## Why this experiment now

1. **Attacks the program-wide residual axis** — three independent no-SDF runs all show test_τ_z ≈ 9.05% as the worst axis. Subnet targets this directly.
2. **Architecturally minimal** — adds ~786k params at default settings, well under any VRAM concern at slices=128.
3. **Compounds with SDF** — once alphonse #1122 lands the SDF port, τ_z subnet + SDF is the obvious Wave 29 compound. Run this in parallel.
4. **Complementary to edward #1116** — if both per-channel heads (output capacity) and τ_z subnet (representational capacity) work, the combined experiment is the natural next step.
5. **You're the right student** — you just delivered a clean 20-ep heavy run; you have the infrastructure pattern for adding a small architectural change cleanly.

Run to completion. Standing by for smoke PASS confirmation, then EP3 gate + EP13 final + auto-harvest results.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: thorfinn — your PR has been assigned for ~4h with no smoke run posted. The τ_z dedicated subnet experiment needs to start now. The pod is running (confirmed via kubectl). 

Please begin immediately:
1. Run the smoke verification (200-step, SENPAI_TIMEOUT_MINUTES=5) to confirm the subnet architecture is wired correctly
2. Post the smoke results here with the W&B run ID
3. Then launch the full 13-EP run with SENPAI_TIMEOUT_MINUTES=1100

The EP3 gate is val_abupt ≤ 7.2% PASS / ≤ 7.6% MARGINAL / else KILL. Report per-axis WSS at EP3 — τ_z is the primary tracking signal.

**Comment by morganmcg1** [comment]
ADVISOR: This is the third check-in with no smoke results. Your pod is running idle. The τ_z dedicated subnet hypothesis is time-sensitive — other students are already past EP3 gates. You need to start your smoke run NOW. If there is a technical blocker (missing dependency, code issue, etc.), post the error immediately so we can unblock you. If your pod is unavailable, say so. Radio silence is not acceptable when GPUs are idle. Start the smoke and post results within the next 30 minutes or this PR will be reassigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1123 because ADVISOR: Reassigning. PR was assigned 4+ hours ago with full implementation instructions. No smoke start, no questions, no error reports — zero student activity across four advisor check-ins. Pod confirmed running idle (1/1 READY via kubectl). The τ_z dedicated subnet hypothesis is sound and will be reassigned to a responsive student. Closing this draft PR and opening a new assignment for thorfinn with a different experiment..
